### ISSUE TITLE
Release in GitLab fails due to can't push to master from detached state

### ISSUE THREAD

**Issue by jwsomis** [projen/projen#1873]
Projen version: 0.56.13
CDK Version: 2.24.1
Typescript version: 4.6.4

I've recently upgraded to the newest version of projen from 0.19.x. I use it to define CDK constructs in a GitLab monorepo where the root project and .gitlab-ci.yml are manually created (not via projen). I used to use the bump task to update version, then manually tag and commit changes back to the repo, but am trying to switch to using projen release command. I'm running into an issue when the release task tries to commit back to the repo where it throws the error
```
👾 release » publish:git | git push --follow-tags origin master
error: src refspec master does not match any
error: failed to push some refs to 'https://gitlab.com/Eturi/devops/cdkconstructs'
```
It does look like the commit is being created, so it's not that there's no commit, but it's in a detached state.
```
[detached HEAD dc08d87] chore(release): 0.1.1
 1 file changed, 2 insertions(+)
```
It is my understanding that due to being in a detached state, I can't just push to master with `git push --follow-tags origin master` which is what the default gitPushCommand is, but rather need `git push --follow-tags origin HEAD:master`. I have seen that you can override the publisher.gitPushCommand, but have not figured out how to do this with the AwsCdkConstructLibrary class.
I can't edit the task with `project.release.publish.publishToGit(gitPushCommand: `git push --follow-tags origin HEAD:master`})`. It tries to create a new task, but one with that name already exists. If I try to first remove the publish:git task, it fails because the release task is dependent on it.

So 1, why does the gitPushCommand try to push directly to the main|master branch if it's in a detached state? Won't that always error?
And 2, how do configure the override for the gitPushCommand? I can't find any docs to help.

Thanks

**Comment by Chriscbr** [comment]
Is it possible to customize the git push command by specifying the release trigger like this perhaps?

```ts
const project = new CustomProject({
  ...
  releaseTrigger: ReleaseTrigger.manual({
    gitPushCommand: "command",
  }),
});
```

I'm not too familiar with this component. If changing `master` to `HEAD:master` wouldn't break any existing users, then I'd be okay accepting a pull request that updates the default command directly in projen's source code.

cc @gradybarrett

**Comment by gradybarrett** [comment]
Hmmmm, the only way that commit would be detached is if you were already in a detached state prior to running the manual release. So what is the gitlab workflow that you have look like? Does it check out a specific commit?

**Comment by jwsomis** [comment]
Thanks for the responses. The help is really appreciated!

@Chriscbr This does work for me, so thank you for this. It would be useful we could pass a function to gitPushCommand so that we could return a template string to allow variables like git branch to be used at runtime (not sure if there is some other way to do this), but the hardcoded string works for me for now.

I'm not certain, but I do believe you could always use HEAD:<branch-name> since HEAD always exists.

@gradybarrett I do not explicitly check out a commit, but it looks like that is the default for gitlab runner per https://gitlab.com/gitlab-org/gitlab/-/issues/15409. 

My current git/npm steps are 
```
before_script:
    - npm config set @eturi:registry https://gitlab.com/api/v4/projects/${CI_PROJECT_ID}/packages/npm/
    - npm config set '//gitlab.com/api/v4/projects/${CI_PROJECT_ID}/packages/npm/:_authToken' "${CI_JOB_TOKEN}"
    - git config --global user.email "<email address>"
    - git config --global user.name "Gitlab CI"
    - git config --global --add safe.directory $(pwd)
    - git remote set-url origin https://oauth2:${ACCESS_TOKEN}@${CI_PROJECT_URL:8}
  ```
The issue above does contain a passable workaround or an ugly, but more complete one. May be worth either documenting for GitLab users or making the `HEAD:master` change

**Comment by jwsomis** [comment]
One other thing, if you're able to answer here. I haven't run into this yet, but since I have a monorepo where CI release steps are run in each constructs directory. I suspect that if I have releases triggered for multiple constructs at once I could hit a race condition where the first construct to push changes back wins and subsequent ones fail. Previously, when I did all these steps manually did `pull --rebase` right before the push to make sure I had the latest. It is a bit weird, but is there anyway to insert this task via projenrc.js in the release steps?

**Comment by gradybarrett** [comment]
Ahhh, gotcha. Does seem overly complex that gitlab is checking out a detached commit. I mean, I get it if all you're doing is building and releasing, but it's not great if you're trying to do what we're doing here. I checked a few circleci checkout tasks and github's checkout to see what other CIs do and they all seem to check out a branch.

Have you looked at the [GitLab model](https://github.com/projen/projen/blob/main/src/gitlab)? @Chriscbr do you know how far along and usable that is? I feel like that might be a better fit, where you can declare your whole workaround above. The manual trigger was only really meant as a rudimentary manual (user-triggered) way for folks to trigger a release - folks not using CI, or those who have very limited to no control over their CI.

Having said that I'm not opposed to HEAD:<branch>. Mostly just trying to understand if it actually makes sense. I wouldn't want to have it introduced, and in doing so allow folks to get themselves into sticky situations - maybe not realizing their in a detached state locally while performing a manual release.

**Comment by pflorek** [comment]
Yes, I had the same problem with GitLab, having a scheduled pipeline updating the dependencies like Github Actions does. I came up with a fragment like this:

```typescript
const { awscdk, gitlab } = require("projen");


const GitlabStages = {
  UPGRADE: "upgrade",
};

const stages = Object.values(GitlabStages);

new gitlab.GitlabConfiguration(project, {
  stages: stages,
  jobs: {
    upgrade: {
      image: { name: "node:16-alpine3.15" },
      stage: GitlabStages.UPGRADE,
      variables: {
        UPGRADE_USER: "$UPGRADE_USER"
        UPGRADE_TOKEN: "$UPGRADE_TOKEN",
      },
      beforeScript: ["apk add --update git bash openssh rsync"],
      script: [
        "git remote set-url origin https://${UPGRADE_USER}:${UPGRADE_TOKEN}@${CI_SERVER_HOST}/${CI_PROJECT_NAMESPACE}/${CI_PROJECT_NAME}.git",
        "yarn install",
        "yarn run upgrade",
        'CHANGED="$(git diff-index --name-only HEAD --)"',
        '[ -s "$CHANGED" ] && exit 0',
        'git config user.name "Patrick Florek"',
        'git config user.email "patrick.florek@gmail.com"',
        "git status",
        "git add .",
        'git commit -m "chore(deps): upgrade main" || exit 0',
        "git push --push-option=ci.skip origin HEAD:$CI_COMMIT_REF_NAME",
      ],
      only: { refs: ["main"] },
    },
  },
});
```
